[FixIrreducible] Use CycleInfo instead of a custom SCC traversal

ssahasra opened the pull request:
[FixIrreducible] Use CycleInfo instead of a custom SCC traversal

1. CycleInfo efficiently locates all cycles in a single pass, while the SCC is
   repeated inside every natural loop.

2. CycleInfo provides a hierarchy of irreducible cycles, and the new
   implementation transforms each cycle in this hierarchy separately instead of
   reducing an entire irreducible SCC in a single step. This reduces the number
   of control-flow paths that pass through the header of each newly created
   loop. This is evidenced by the reduced number of predecessors on the "guard"
   blocks in the lit tests, and fewer operands on the corresponding PHI nodes.

3. When an entry of an irreducible cycle is the header of a child natural loop,
   the original implementation destroyed that loop. This is now preserved,
   since the incoming edges on non-header entries are not touched.

4. In the new implementation, if an irreducible cycle is a superset of a natural
   loop with the same header, then that natural loop is destroyed and replaced
   by the newly created loop.

ssahasra commented:
Note that I have not yet finished verifying all the lit tests. I might also have to add a few more tests, especially involving a mix of irreducible and reducible cycles that are siblings and/or nested inside each other in various combinations. Especially with some overlap in the entry and header nodes.

kuhar commented on llvm/include/llvm/ADT/GenericCycleInfo.h:
```diff
@@ -189,6 +195,21 @@ template <typename ContextT> class GenericCycle {
   using const_entry_iterator =
       typename SmallVectorImpl<BlockT *>::const_iterator;
+  const_entry_iterator entry_begin() const {
+    return const_entry_iterator{Entries.begin()};
```
nit: you should be able to drop the redundant return types here and below
```suggestion
    return {Entries.begin()};
```

ssahasra replied:
Thanks! Fixed in an intermediate patch.

ssahasra replied:
Fixed.

kuhar commented on llvm/include/llvm/ADT/GenericCycleInfo.h:
```diff
@@ -107,6 +107,12 @@ template <typename ContextT> class GenericCycle {
+  /// \brief Replace all entries with \p Block as single entry.
+  void setSingleEntry(BlockT *Block) {
+    Entries.clear();
+    Entries.push_back(Block);
```
Could this be `nullptr`? Should we add an assert to guard against this?

ssahasra replied:
Good catch! Replaced it with a stronger test to check that `Block` is contained in the current cycle.

ssahasra replied:
Fixed.

ssahasra commented:
Current status: This is going to take a "bit" longer. Turns out that `CreateControlFlowHub` is a heavy hammer which redirects all edges from `Incoming` to `Outgoing`. But this destroys inner cycles when used to fix an irreducible cycle. For some cycle `C` with a subcycle `S`, a block `B` in `S` may have edges going to the header `H` of `C` as well as the header `E` of `S`. Now if the header `E` is also an entry of `C`, then the use of `CreateControlFlowHub` removes the edge from `B` to `E`, thus destroying `S`.

This needs a finer method that redirects only specific edges. Either that, or we let the pass destroy some cycles. But updating `CycleInfo` for these missing subcycles may be a fair amount of work too, so I would rather do it the right way.

ssahasra commented:
> Note that I have not yet finished verifying all the lit tests. I might also have to add a few more tests, especially involving a mix of irreducible and reducible cycles that are siblings and/or nested inside each other in various combinations. Especially with some overlap in the entry and header nodes.

- New tests added that involve nesting with common header or entry nodes. Existing tests also covered some relevant combinations.
- Verified all tests.

ssahasra committed:
[FixIrreducible] Use CycleInfo instead of a custom SCC traversal

1. CycleInfo efficiently locates all cycles in a single pass, while the SCC is
   repeated inside every natural loop.

2. CycleInfo provides a hierarchy of irreducible cycles, and the new
   implementation transforms each cycle in this hierarchy separately instead of
   reducing an entire irreducible SCC in a single step. This reduces the number
   of control-flow paths that pass through the header of each newly created
   loop. This is evidenced by the reduced number of predecessors on the "guard"
   blocks in the lit tests, and fewer operands on the corresponding PHI nodes.

3. When an entry of an irreducible cycle is the header of a child natural loop,
   the original implementation destroyed that loop. This is now preserved,
   since the incoming edges on non-header entries are not touched.

4. In the new implementation, if an irreducible cycle is a superset of a natural
   loop with the same header, then that natural loop is destroyed and replaced
   by the newly created loop.
